source/common/http/conn_manager_impl.cc

```diff
@@ -1398,7 +1397,7 @@ void ConnectionManagerImpl::ActiveStream::decodeHeaders(RequestHeaderMapSharedPt
   // should return 404. The current returns no response if there is no router filter.
   if (hasCachedRoute()) {
     // Do not allow upgrades if the route does not support it.
-    if (upgrade_rejected) {
+    if (state_.upgrade_rejected_) {
```

Comment:
It's my initial design and I created `CreateFilterChainResult` struct. See https://github.com/envoyproxy/envoy/pull/37112#discussion_r1842746304

But after I found the `upgradeFilterChainCreated` callback, I think may be use the callback to report the upgrade result in a uniform way would be better. It will make the code easy to read or understand if you read a from scrach.

By this way, the `createFilterChain` method only return a bool to tell if the filter chain is created or not. This is also more friendly to upstream filter chain becase the upstream filter chain doesn't care upgrade.

If we inclined to use the return value. Then I will suggest to remove  `upgradeFilterChainCreated` callback and use the return value to report all results: filter chain created, upgrade accepted, upgrade rejected. (if both upgrade accepted and upgrade rejected are false, then means no upgrade).

In conlusion, I prefer to use same way to report the upgrade result, no matter return value or callback.

Reply:
yeah I like removing the callback and doing it all inline - IMO it makes the code path simpler an easier to read
/wait